<a href="https://colab.research.google.com/github/Sreenathkk00/Machine-Learning-Projects/blob/main/Purchase_Prediction_And_Micronumerosity.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Purchase Prediction And Micronumerosity**

In [100]:
# Import Libraries
import pandas as pd
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score,confusion_matrix,classification_report
from sklearn.ensemble import RandomForestClassifier

In [101]:
# Import Data
purchase = pd.read_csv('https://github.com/YBI-Foundation/Dataset/raw/main/Customer%20Purchase.csv')

Display Data Information

In [102]:
purchase.head()

Customer ID  Age  Gender Education   Review Purchased
0         1021   30  Female    School  Average        No
1         1022   68  Female        UG     Poor        No
2         1023   70  Female        PG     Good        No
3         1024   72  Female        PG     Good        No
4         1025   16  Female        UG  Average        No

In [103]:
purchase.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50 entries, 0 to 49
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Customer ID  50 non-null     int64 
 1   Age          50 non-null     int64 
 2   Gender       50 non-null     object
 3   Education    50 non-null     object
 4   Review       50 non-null     object
 5   Purchased    50 non-null     object
dtypes: int64(2), object(4)
memory usage: 2.5+ KB


In [104]:
purchase.describe()

Customer ID        Age
count     50.00000  50.000000
mean    1045.50000  54.160000
std       14.57738  25.658161
min     1021.00000  15.000000
25%     1033.25000  30.250000
50%     1045.50000  57.000000
75%     1057.75000  74.000000
max     1070.00000  98.000000

Data Preprocessing

In [105]:
# Initialize LabelEncoder
label_encoders = {}
# Encode categorical columns
for column in ['Gender', 'Education', 'Review', 'Purchased']:
    le = LabelEncoder()
    purchase[column] = le.fit_transform(purchase[column])
    label_encoders[column] = le
print(purchase.head())

   Customer ID  Age  Gender  Education  Review  Purchased
0         1021   30       0          1       0          0
1         1022   68       0          2       2          0
2         1023   70       0          0       1          0
3         1024   72       0          0       1          0
4         1025   16       0          2       0          0


In [106]:
# defind target(y) and feature(x)
y = purchase['Purchased']
x = purchase.drop('Purchased',axis=1)

In [107]:
# Train-Test model
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.3,random_state=2529)

In [108]:
#Shape of the test train models
x_train.shape,x_test.shape,y_train.shape,y_test.shape

((35, 5), (15, 5), (35,), (15,))

Model 1- LogisticRegression

In [109]:
# model selection
from sklearn.linear_model import LogisticRegression
model = LogisticRegression()

In [110]:
# Train the model
model.fit(x_train,y_train)


LogisticRegression()

In [111]:
# Make Predictions
y_pred = model.predict(x_test)
y_pred

array([0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 1])

Model Evaluation

In [112]:
# Accuracy score
accuracy_score(y_test,y_pred)


0.26666666666666666

In [113]:
# TP,TP & FN,FP
confusion_matrix(y_test,y_pred)

array([[3, 4],
       [7, 1]])

In [114]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.30      0.43      0.35         7
           1       0.20      0.12      0.15         8

    accuracy                           0.27        15
   macro avg       0.25      0.28      0.25        15
weighted avg       0.25      0.27      0.25        15



model 2 : RandomForestClassifier

In [115]:
# select the model
model = RandomForestClassifier(random_state=42)

In [116]:
# Train the model
model.fit(x_train,y_train)

RandomForestClassifier(random_state=42)

In [117]:
# make prediction
y_pred = model.predict(x_test)
y_pred

array([0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0])

In [118]:
confusion_matrix(y_test,y_pred)

array([[7, 0],
       [7, 1]])

In [119]:
accuracy_score(y_test,y_pred)

0.5333333333333333

In [120]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.50      1.00      0.67         7
           1       1.00      0.12      0.22         8

    accuracy                           0.53        15
   macro avg       0.75      0.56      0.44        15
weighted avg       0.77      0.53      0.43        15



# Customer Purchase Prediction

This project aims to predict whether a customer will make a purchase based on various features such as age, gender, education, and review. Two machine learning models, Logistic Regression and Random Forest Classifier, were implemented and evaluated.

## Project Structure

- **Data Import and Exploration**: Loading the dataset and exploring its structure.
- **Data Preprocessing**: Encoding categorical variables and splitting the data into training and testing sets.
- **Model Training and Evaluation**:
  - Logistic Regression
  - Random Forest Classifier
- **Model Performance**: Evaluation using accuracy score, confusion matrix, and classification report.

## Results

### Logistic Regression
- **Accuracy**: 0.27
### Random Forest Classifier
- **Accuracy**: 0.53


